packages/typespec-ts/static/static-helpers/uriTemplate.ts

```diff
@@ -0,0 +1,169 @@
+    return value !== undefined && value !== null;
+}
+
+function isKeyOperator(operator: string): boolean {
+    return operator === ";" || operator === "&" || operator === "?";
+}
+
+function getValues(
+    context: any,
+    operator: any,
+    key: any | null,
+    modifier: any | null,
+) {
+    var value = context[key],
+        result: any[] = [];
+
+    if (isDefined(value) && value !== "") {
+        if (
+            typeof value === "string" ||
+            typeof value === "number" ||
+            typeof value === "boolean"
+        ) {
+            value = value.toString();
+
+            if (modifier && modifier !== "*") {
+                value = value.substring(0, parseInt(modifier, 10));
+            }
+            result.push(
+                encodeValue(operator, value, isKeyOperator(operator) ? key : null),
+            );
+        } else {
+            if (modifier === "*") {
+                if (Array.isArray(value)) {
+                    value.filter(isDefined).forEach(function (value) {
+                        result.push(
+                            encodeValue(
+                                operator,
+                                value,
+                                isKeyOperator(operator) ? key : null,
+                            ),
+                        );
+                    });
+                } else {
+                    Object.keys(value).forEach(function (k) {
+                        if (isDefined(value[k])) {
+                            result.push(encodeValue(operator, value[k], k));
+                        }
+                    });
+                }
+            } else {
+                var tmp: any[] = [];
+
+                if (Array.isArray(value)) {
+                    value.filter(isDefined).forEach(function (value) {
+                        tmp.push(encodeValue(operator, value));
+                    });
+                } else {
+                    Object.keys(value).forEach(function (k) {
+                        if (isDefined(value[k])) {
+                            tmp.push(encodeUnreserved(k));
+                            tmp.push(encodeValue(operator, value[k].toString()));
+                        }
+                    });
+                }
+
+                if (isKeyOperator(operator)) {
+                    result.push(encodeUnreserved(key) + "=" + tmp.join(","));
+                } else if (tmp.length !== 0) {
+                    result.push(tmp.join(","));
+                }
+            }
+        }
+    } else {
+        if (operator === ";") {
+            if (isDefined(value)) {
+                result.push(encodeUnreserved(key));
+            }
+        } else if (value === "" && (operator === "&" || operator === "?")) {
+            result.push(encodeUnreserved(key) + "=");
+        } else if (value === "") {
+            result.push("");
+        }
+    }
+    return result;
+}
+
+export function parseTemplate(template: any) {
```

Comment:
@joheredi I verified we could leverage a 3rd party lib
[url-template](https://www.npmjs.com/package/url-template) to support full uri template cases in modular.
One thing to confirm with you is if we need to re-implement a url-template by ourselves. 

My preference is Yes, main are two considerations. The first is for copyright, that lib is under personal copy right. That would be a concern for us I believe but you could share more with me. https://github.com/bramstein/url-template/blob/main/LICENSE
 
The second is for controllability, we could fully control any bug fixing or new features if we implement in our side. For example we may have feature to add skipUrlEncoding option in option bag and this would lead any change for that lib. Also this logic is quite simple and they are just around 100 lines.

Any concerns?

Reply:
@xirzec how do you think?

Reply:
I agree that's it's not worth a runtime dependency for ~100 lines, especially since we've had to deal with Azure specific quirks in url templates before